File: extensions/youtube/src/main/java/app/morphe/extension/youtube/patches/ExpandLivestreamDvrDurationPatch.java
package app.morphe.extension.youtube.patches;

import app.morphe.extension.youtube.settings.Settings;

@SuppressWarnings("unused")
public class ExpandLivestreamDvrDurationPatch {

    private static final double SEVEN_DAYS_SEC = 7.0 * 24 * 60 * 60;

    /**
     * Injection point.
     */
    public static double overrideMaxDvrDurationSec(double originalDurationSec) {
        if (!Settings.EXPAND_LIVESTREAM_DVR_DURATION.get()) return originalDurationSec;
        if (originalDurationSec <= 0) return originalDurationSec;
        return SEVEN_DAYS_SEC;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/youtube/interaction/seekbar/ExpandLivestreamDvrDurationPatch.kt
package app.morphe.patches.youtube.interaction.seekbar

import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.extensions.InstructionExtensions.getInstruction
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.shared.misc.settings.preference.SwitchPreference
import app.morphe.patches.youtube.misc.extension.sharedExtensionPatch
import app.morphe.patches.youtube.misc.settings.PreferenceScreen
import app.morphe.patches.youtube.misc.settings.settingsPatch
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction

private const val EXTENSION_CLASS_DESCRIPTOR =
    "Lapp/morphe/extension/youtube/patches/ExpandLivestreamDvrDurationPatch;"

val expandLivestreamDvrDurationPatch = bytecodePatch(
    description = "Expands the seekable duration of livestream DVR to 7 days.",
) {
    dependsOn(
        sharedExtensionPatch,
        settingsPatch,
    )

    execute {
        PreferenceScreen.SEEKBAR.addPreferences(
            SwitchPreference("morphe_expand_livestream_dvr_duration")
        )

        FormatStreamModelMaxDvrDurationFingerprint.method.apply {
            val returnIndex = FormatStreamModelMaxDvrDurationFingerprint.instructionMatches.last().index
            val returnReg = getInstruction<OneRegisterInstruction>(returnIndex).registerA

            addInstructions(
                returnIndex,
                """
                    invoke-static { v$returnReg, v${returnReg + 1} }, $EXTENSION_CLASS_DESCRIPTOR->overrideMaxDvrDurationSec(D)D
                    move-result-wide v$returnReg
                """
            )
        }
    }
}
